src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs

```diff
@@ -0,0 +1,385 @@
+        let input = self.visit(plan.input());
+
+        if plan.group_key().is_empty() {
+            input.limit_to(1)
+        } else {
+            input.as_low_as(1)
+        }
+    }
+
+    fn visit_logical_limit(&mut self, plan: &plan_node::LogicalLimit) -> Cardinality {
+        self.visit(plan.input()).limit_to(plan.limit() as usize)
+    }
+
+    fn visit_logical_project(&mut self, plan: &plan_node::LogicalProject) -> Cardinality {
+        self.visit(plan.input())
+    }
+
+    fn visit_logical_top_n(&mut self, plan: &plan_node::LogicalTopN) -> Cardinality {
+        let input = self.visit(plan.input());
+
+        match plan.limit_attr() {
+            Limit::Simple(limit) => input.sub(plan.offset() as usize).limit_to(limit as usize),
+            Limit::WithTies(limit) => {
+                assert_eq!(plan.offset(), 0, "ties with offset is not supported yet");
+                input.as_low_as(limit as usize)
+            }
+        }
+    }
+
+    fn visit_logical_filter(&mut self, plan: &plan_node::LogicalFilter) -> Cardinality {
+        let input = plan.input();
+        let eq_set: HashSet<_> = plan
+            .predicate()
+            .collect_input_refs(input.schema().len())
+            .ones()
+            .collect();
+
+        let mut unique_keys: Vec<HashSet<_>> = vec![input.logical_pk().iter().copied().collect()];
+
+        // We don't have UNIQUE key now. So we hack here to support some complex queries on
+        // system tables.
+        if let Some(scan) = input.as_logical_scan() && scan.is_sys_table() && scan.table_name() == PG_NAMESPACE_TABLE_NAME {
+            let nspname = scan.output_col_idx().iter().find(|i| scan.table_desc().columns[**i].name == "nspname").unwrap();
+            unique_keys.push([*nspname].into_iter().collect());
+        }
+
+        let input = self.visit(input);
+        if unique_keys
+            .iter()
+            .any(|unique_key| eq_set.is_superset(unique_key))
+        {
+            input.limit_to(1).as_low_as(0)
+        } else {
+            input.as_low_as(0)
+        }
+    }
+
+    fn visit_logical_union(&mut self, plan: &plan_node::LogicalUnion) -> Cardinality {
+        if plan.all() {
+            plan.inputs()
+                .into_iter()
+                .map(|input| self.visit(input))
+                .reduce(std::ops::Add::add)
+                .unwrap_or_default()
+        } else {
+            Cardinality::default()
+        }
+    }
+
+    fn visit_logical_join(&mut self, plan: &plan_node::LogicalJoin) -> Cardinality {
+        let left = self.visit(plan.left());
+        let right = self.visit(plan.right());
+
+        match plan.join_type() {
+            JoinType::Unspecified => unreachable!(),
+
+            JoinType::Inner => left.mul(right).as_low_as(0),
+
+            JoinType::LeftOuter | JoinType::LeftSemi | JoinType::LeftAnti => left.as_low_as(0),
+            JoinType::RightOuter | JoinType::RightSemi | JoinType::RightAnti => right.as_low_as(0),
```

Comment:
The left outer join will only increase the number of output rows on the left side, it will not decrease them. The same principle applies to the right outer join.

Reply:
Makes sense.